sqldb/sqlite.go

```diff
@@ -0,0 +1,176 @@
+package sqldb
+
+import (
+	"database/sql"
+	"fmt"
+	"net/url"
+	"path/filepath"
+	"testing"
+	"time"
+
+	sqlite_migrate "github.com/golang-migrate/migrate/v4/database/sqlite"
+	"github.com/lightningnetwork/lnd/sqldb/sqlc"
+	"github.com/stretchr/testify/require"
+	_ "modernc.org/sqlite" // Register relevant drivers.
+)
+
+const (
+	// sqliteOptionPrefix is the string prefix sqlite uses to set various
+	// options. This is used in the following format:
+	//   * sqliteOptionPrefix || option_name = option_value.
+	sqliteOptionPrefix = "_pragma"
+
+	// sqliteTxLockImmediate is a dsn option used to ensure that write
+	// transactions are started immediately.
+	sqliteTxLockImmediate = "_txlock=immediate"
+
+	// defaultMaxConns is the number of permitted active and idle
+	// connections. We want to limit this so it isn't unlimited. We use the
+	// same value for the number of idle connections as, this can speed up
+	// queries given a new connection doesn't need to be established each
+	// time.
+	defaultMaxConns = 25
+
+	// connIdleLifetime is the amount of time a connection can be idle.
+	connIdleLifetime = 5 * time.Minute
+)
+
+// SqliteConfig holds all the config arguments needed to interact with our
+// sqlite DB.
+//
+//nolint:lll
+type SqliteConfig struct {
+	// SkipMigrations if true, then all the tables will be created on start
+	// up if they don't already exist.
+	SkipMigrations bool `long:"skipmigrations" description:"Skip applying migrations on startup."`
+
+	// DatabaseFileName is the full file path where the database file can be
+	// found.
+	DatabaseFileName string `long:"dbfile" description:"The full path to the database."`
+}
+
+// SqliteStore is a database store implementation that uses a sqlite backend.
+type SqliteStore struct {
+	cfg *SqliteConfig
+
+	*BaseDB
+}
+
+// NewSqliteStore attempts to open a new sqlite database based on the passed
+// config.
+func NewSqliteStore(cfg *SqliteConfig) (*SqliteStore, error) {
+	// The set of pragma options are accepted using query options. For now
+	// we only want to ensure that foreign key constraints are properly
+	// enforced.
+	pragmaOptions := []struct {
+		name  string
+		value string
+	}{
+		{
+			name:  "foreign_keys",
+			value: "on",
+		},
+		{
+			name:  "journal_mode",
+			value: "WAL",
+		},
+		{
+			name:  "busy_timeout",
+			value: "5000",
+		},
+		{
+			// With the WAL mode, this ensures that we also do an
+			// extra WAL sync after each transaction. The normal
+			// sync mode skips this and gives better performance,
+			// but risks durability.
+			name:  "synchronous",
+			value: "full",
+		},
+		{
+			// This is used to ensure proper durability for users
```

Comment:
While this is true, iirc mac users will suffer greatly when it comes to performance. Some itests may even fail too. I think the mac philosophy is that writes are durable as long as there's battery remaining and even then the OS is capable to putting the system to sleep when needed.

Reply:
FWIW this is only active on mac. It _is_ the case that this dramatically improves perf if removed. IIRC, for `sqlite` elsewhere in the project, we move these directives up one level and allow users to override them.